model/lifecycle.go

```diff
@@ -159,20 +159,32 @@ func MarkVersionCompleted(versionId string, finishTime time.Time) error {
 
 	// Find the statuses for all builds in the version so we can figure out the version's status
 	builds, err := build.Find(
-		build.ByVersion(versionId).WithFields(build.StatusKey),
+		build.ByVersion(versionId).WithFields(build.StatusKey, build.TasksKey),
 	)
 	if err != nil {
 		return err
 	}
 
+	allCachedTasksInActiveBuildsComplete := true
+	finished := true
 	for _, b := range builds {
 		if !b.IsFinished() {
-			return nil
+			finished = false
+			continue
+		}
+		if !b.AllCachedTasksOrCompileFinished() {
+			allCachedTasksInActiveBuildsComplete = false
 		}
 		if b.Status != evergreen.BuildSucceeded {
 			status = evergreen.VersionFailed
 		}
 	}
+	if allCachedTasksInActiveBuildsComplete {
+		event.LogVersionStateChangeEvent(versionId, status)
+	}
+	if !finished {
```

Comment:
It feels like `allCachedTasksInActiveBuildsComplete` should do the same thing as `finished`. Is your goal here to log a version state change event in some cases where you do not update the version status?

Reply:
Suppose a version has 2 builds. Build A is scheduled and succeeded, but Build B was never scheduled. Examples of these might be aborting a build in a version (oops I made a mistake), or another example is this evergreen version: https://evergreen.mongodb.com/version/mci_6c0bbbd4170e56aafc90c534588fc50595123cf3, which features a bunch of builds that have run to completion, and a bunch that have never been run.

In both cases, a notification should be sent, but the problem is if you look at the version's status in the database, it's still VersionStarted. If you look at this code, it's because finished != allCachedTasksInActiveBuildsComplete in all cases, specifically when builds are inactive/unscheduled. 

To fix this properly (like I was trying to do last week) breaks things in a more far reaching way, that is very risk and very difficult to debug.